Fix rebalancing with rounding errors

hunchr opened the pull request:
TICKET-19655
Rebalancing can produce unfixable mismatch between timer session and time entries.
For example if you start a timer session with 1h on three tickets, then clicking "Ausbalancieren" will not work.

* Fix these rounding errors

schmijos commented on app/services/time_discrepancy_loader.rb:
```diff
@@ -1,26 +1,22 @@
+      .group('timer_sessions.id')
+      .having(
+        <<~SQL
+          timer_sessions.hours != SUM(time_entries.hours)
+          OR timer_sessions.hours != #{time_difference_expr} / 3600
```
Why do we need both conditions here?

hunchr replied:
Because we're comparing 3 dates: the time diff between start and end, the summed up hours of each entry and the total hours

schmijos replied:
Ok, I checked this. This seems fine.

schmijos commented on app/controllers/timer_sessions_controller.rb:
```diff
@@ -5,15 +5,8 @@ def index
-                      .reverse
-                      .group_by(&:entry_date)
+    @non_matching_timer_session_ids = TimeDiscrepancyLoader.uneven_timer_sessions(@timer_sessions_in_range).map(&:id)
+    timer_sessions
```
Please express what you're doing with the name of the method. `set_timer_sessions` is better.

hunchr replied:
I haven't renamed any methods

hunchr replied:
Shouldn't getters and setters avoid using set_ in ruby?

schmijos replied:
You're mixing up two things:

* Setting instance values from outside the class:
  We rather use setters in Ruby like this
  ```rb
  def timer_sessions=(value)
    @whatever = value
  end
  ```
  than like this
  ```rb
  def set_timer_sessions(value)
    @whatever = value
  end
  ```
* Setting instance values from inside the class:
  Calling `timer_sessions` just like this doesn't expose any instance altering behaviour to the reader,
  but something like `set_timer_sessions`, `configure_timer_sessions` or `setup_timer_sessions` would.

schmijos reviewed (changes requested):
Tests are missing. Please implement the regression from the ticket description.

hunchr commented:
> Tests are missing. Please implement the regression from the ticket description.

I only refactored the code so there isn't anything new to test

schmijos reviewed (changes requested):
I disagree. This is not a refactoring. You are altering the behaviour of the code. So you need to write a regression test. Even your PR is called "Fix …", so I actually hope your code is not only a refactoring, because a refactoring wouldn't fix any bugs.

hunchr commented on app/services/time_discrepancy_loader.rb:
```diff
@@ -5,22 +5,19 @@ class TimeDiscrepancyLoader
-             AS DECIMAL(10, #{DECIMALS_TO_ROUND_TO}))
+             timer_sessions.hours != CAST(SUM(time_entries.hours) AS DECIMAL(10, #{DECIMALS_TO_ROUND_TO}))
+             OR timer_sessions.hours != #{time_difference_expr} / 3600
            SQL
```
Fixed the SQL query

hunchr commented on test/factories/timer_session.rb:
```diff
@@ -10,19 +10,22 @@
+        3.times do
+          e = TimeEntry.create(user: User.current, project: Project.last, hours: 1.0 / 3, spent_on: Time.zone.today)
+          TimerSessionTimeEntry.create!(timer_session:, time_entry: e)
+        end
```
add factory trait

hunchr commented on test/unit/time_discrepancy_loader_test.rb:
```diff
@@ -45,4 +45,13 @@ class TimeDiscrepancyLoaderTest < ActiveSupport::TestCase
+    non_matching_timer_session_ids = TimeDiscrepancyLoader.uneven_timer_sessions(timer_sessions_in_range).map(&:id)
+
+    assert non_matching_timer_session_ids.exclude?(timer_session.id)
+  end
```
add test

hunchr replied:
add tests

hunchr commented:
> So you need to write a regression test

I wrote a test and also pulled from main because the CI was failing. now it should work

schmijos commented on PluginGemfile:
```diff
@@ -3,9 +3,9 @@ source 'https://rubygems.org'
-  gem 'bullet'
-  gem 'factory_bot_rails'
+  gem 'bullet', '~> 6.1'
+  gem 'factory_bot_rails', '~> 6.4'
   gem 'rubocop', '~> 1.26.0'
   gem 'rubocop-performance', '~> 1.13.0'
-  gem 'brakeman'
+  gem 'brakeman', '~> 6.0'
```
Why do you want to lock them so tightly?

hunchr replied:
that is not part of this PR, it's just a merge

schmijos commented on test/functional/completion_controller_test.rb:
Nice! Thank you!

schmijos reviewed (changes requested):
This PR contains too many things. Please rebase ontop of the other changes, so that I can review it more comfortably.

<img width="1583" alt="CleanShot 2023-12-11 at 17 28 42@2x" src="https://github.com/renuo/redmine_tracky/assets/245443/edffde61-04ec-4efd-bd1e-b96d0d524a30">

hunchr commented:
I rebased the commits. Now you only need to review the first one.
<img width="518" alt="image" src="https://github.com/renuo/redmine_tracky/assets/76159444/43467fa9-6ede-44d3-b79e-52721cb8e92e">

hunchr commented on test/factories/timer_session.rb:
```diff
@@ -10,19 +10,22 @@
-          timer_session: timer_session,
-          time_entry: TimeEntry.first
-        )
+        TimerSessionTimeEntry.create!(timer_session:, time_entry: TimeEntry.first)
```
I converted these traits into one-liners because rubocop complained about the line count

schmijos commented on test/factories/timer_session.rb:
```diff
@@ -10,19 +10,22 @@
+      end
+    end
+
+    trait :with_unrounded_time_entries do
+      after(:create) do |timer_session|
+        3.times do
+          e = TimeEntry.create(user: User.current, project: Project.last, hours: 1.0 / 3, spent_on: Time.zone.today)
+          TimerSessionTimeEntry.create!(timer_session:, time_entry: e)
+        end
       end
     end
```
Please move this to the test. You use it only once. Also I don't think the dependency on the user is wise to have in the factory. It should be agnostic.

schmijos commented on test/unit/time_discrepancy_loader_test.rb:
```diff
@@ -45,4 +45,13 @@ class TimeDiscrepancyLoaderTest < ActiveSupport::TestCase
+                                          .finished.created_by(User.current)
+    non_matching_timer_session_ids = TimeDiscrepancyLoader.uneven_timer_sessions(timer_sessions_in_range).map(&:id)
+
+    assert non_matching_timer_session_ids.exclude?(timer_session.id)
```
In my eyes you don't test for the correct (easy to read thing). What I want to read in this test is in prose:

* Setup timer session for 1.0 hours to be distributed over 3 tickets.
* Call distribution magic
* Expect that there are 3 time entries with a sum of 1.0 hours
* Make assumptions about each time entry, e.g. each has 1/3 hours.

hunchr commented:
I updated the specs but there is not really anything else I could change because I didn't really touch the rebalancing function, I just fixed the SQL query.

hunchr commented:
The CI is failing but that's not because of my changes. This could be fixed in another bugfix branch.

schmijos commented on app/services/time_discrepancy_loader.rb:
```diff
@@ -5,28 +5,25 @@ class TimeDiscrepancyLoader
 
   def self.using_postgresql?
     ActiveRecord::Base.connection.adapter_name == 'PostgreSQL'
+  def self.round(attribute)
+    "CAST(#{attribute} AS DECIMAL(10, #{DECIMALS_TO_ROUND_TO}))"
   end
```
This method is dangerous because it could be called with non-numeric fields containing SQL injectable code. Please make it a private method at least.

schmijos commented on app/services/time_discrepancy_loader.rb:
```diff
@@ -5,28 +5,25 @@ class TimeDiscrepancyLoader
+    "CAST(#{attribute} AS DECIMAL(10, #{DECIMALS_TO_ROUND_TO}))"
   end
+
+  # def self.uneven_timer_session_ids(scope)
+  #   debugger
+  #   return [] if scope.empty?
+
+  #   scope.joins(timer_session_time_entries: :time_entry)
+  #        .group('timer_sessions.id')
+  #        .having("CAST(SUM(time_entries.hours) AS DECIMAL(10, #{DECIMALS_TO_ROUND_TO})) != ?",
+  #                scope.first.hours.round(DECIMALS_TO_ROUND_TO))
+  #        .ids
+  # end
```
Please delete if not used.

schmijos commented on test/unit/time_discrepancy_loader_test.rb:
```diff
@@ -29,9 +30,17 @@ class TimeDiscrepancyLoaderTest < ActiveSupport::TestCase
       timer_session.time_entries.update(hours: timer_session.splittable_hours)
     end
+
+    @timer_session = FactoryBot.create(:timer_session, user: User.current)
+
+    3.times do
+      @timer_session.time_entries.create(user: User.current, project: Project.last,
+                                         hours: 1.0 / 3, spent_on: Time.zone.today)
+    end
```
Where are your beautiful tests?

hunchr commented:
I moved this PR to https://github.com/renuo/redmine_tracky/pull/119
